### PR TITLE
Migrate codebase to concurrent mode, change hotkey lib, fix memory leaks and other minor changes

### LINKED ISSUE #381: Review alternatives for react-hotkeys
Blocker for #378.

See e.g. https://github.com/JohannesKlauss/react-hotkeys-hook.

### PR DESCRIPTION
**Pull request by CodingDive**
Closes https://github.com/edtr-io/edtr-io/issues/381

This got to be a much bigger PR than I initially anticipated and based on feedback, next time I'll split it into multiple smaller PRs. It includes several improvements to our codebase. Key highlights are outlined below:

1. **Migrate to Concurrent Mode**: I have migrated our codebase to use React's Concurrent Mode. StrictMode is now enabled through the `next.config.js`, which should allow us to catch more bugs in development. To accomplish this, we had to migrate from `react-hotkeys` to `react-hotkeys-hooks`. If you run into weird issues when effects fire twice, please reach out to me and we can try to fix the issues together.

2. **Memory Leak Fixes**: In the process of this migration, I found and resolved minor potential memory leaks in our application. This was primarily achieved by ensuring proper handling of timeouts, intervals, fetch requests, and eventListeners. Note that while I have made a considerable effort to address these issues, there may be a few instances I missed. These will be fixed in subsequent PRs.

3. **Code Cleanup and Refactoring**: This PR also allowed for a more extensive cleanup of our code. I took this opportunity to refactor some of our inner `renderComponents` functions to dedicated components in bigger files, improve the typing system, implement early exits where possible, and make some minor style changes.

4. **Small Accessibility win in quickbar**: Now doesn't allow the links to be tabbed which adds a lot to the keyboard experience on our landing page

Sadly, the very issue that made me dig deep into the code base and trying to fix a problem with stale state updates in a Redux action creator in the root of the editor couldn't be fixed with any of the changes I made. I'll try to look into it deeper during my trip to Munich.

Added some new test cases here https://github.com/serlo/frontend-e2e-tests/pull/43

### CODE REVIEW

**Review by Entkenntnis — CHANGES_REQUESTED** [review]
these are too many changes all at once in one PR - this is very difficult to review for me.

Focus! @CodingDive Can you extract the part with react-hotkeys (and this part only) into a separate PR?

**Comment by elbotho** [comment]
@CodingDive did you actually do this in two commits or do you have a branch without the commits squashed?

File: src/serlo-editor/core/sub-document/editor.tsx
Comment (hejtful) [inline src/serlo-editor/core/sub-document/editor.tsx]: I'm curious to know if there is any reason for not moving this hook into its own file? There, it could also have its arguments cleanly defined as an interface and have its own import scope.

**Review by hejtful — APPROVED** [review]
Looks good to me, but it was quite hard to review, and I might have missed something, so I'd like to test it more on staging once it's merged.

Please note that I mainly focused on the editor changes, and I'd recommend waiting for a review from @elbotho or @Entkenntnis before merging.

Edit:
This PR definitely resolves https://github.com/edtr-io/edtr-io/issues/381 , could you put that in the description so that the issue automatically closes when the PR is merged?
It also maybe resolves [this issue](https://github.com/edtr-io/edtr-io/issues/272) with CTRL+Z not working after dragging and dropping a plugin to rearrange them. Would be worth to check it and eventually also add a "resolves" to the description like for 381.

**Comment by CodingDive** [comment]
> @CodingDive did you actually do this in two commits or do you have a branch without the commits squashed?

Sorry, I'm used to squashing and rebase flows from my prior org. I don't have any more commits for this one, but will henceforth always commit my WIP in subsequent PRs.

**Comment by Entkenntnis** [comment]
Again, please split this PR into at least two parts: The `react-hotkeys`-part is very welcomed and I would like to see how a minimal version with `reactStrictMode` looks like.

The code cleanup part is a completely separate issue. It breaks with some consistencies that we have in the frontend (e.g. one one component/export per file) and I like to review and discuss is separately.

Overall, it's just too much happening in the PR for your first bigger PR in this repo. It's ok to take it a bit slower. Again, properly solving https://github.com/edtr-io/edtr-io/issues/381 is already a great achievement.

**Comment by elbotho** [comment]
image
that's what I like to see :)

**Comment by CodingDive** [comment]
Closing this in favor of https://github.com/serlo/frontend/pull/2598. May reopen or cherry pick some things for smaller PRs.

**Pull request closed without merging**
